### PR TITLE
[DependencyInjection] resolve aliases in factories

### PR DESCRIPTION
**Pull request by xabbuh**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | FriendsOfSymfony/FOSUserBundle#2048 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveReferencesToAliasesPass.php [hunk 1]
```diff
@@ -76,6 +78,21 @@ private function processArguments(array $arguments)
         return $arguments;
     }
 
+    private function processFactory($factory)
+    {
+        if (null === $factory || !is_array($factory) || !$factory[0] instanceof Reference) {
+            return;
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveReferencesToAliasesPass.php]: this needs to return the factory

Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveReferencesToAliasesPass.php]: good catch, fixed

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveReferencesToAliasesPass.php [hunk 2]
```diff
@@ -76,6 +78,21 @@ private function processArguments(array $arguments)
...
+    private function processFactory($factory)
+    {
+        if (null === $factory || !is_array($factory) || !$factory[0] instanceof Reference) {
+            return $factory;
+        }
+
+        $defId = $this->getDefinitionId($id = (string) $factory[0]);
+
+        if ($defId !== $id) {
+            $factory[0] = new Reference($defId, $factory[0]->getInvalidBehavior(), $factory[0]->isStrict());
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveReferencesToAliasesPass.php]: note that this line will need changes when merging into 2.8 (and then again in 3.0) due to changes to ``isStrict``

**Comment by stof** [comment]
:+1:

**Comment by stof** [comment]
Status: reviewed

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by fabpot** [comment]
@xabbuh Can I let you merge 2.3 into 2.7, 2.7 into 2.8, and 2.8 into 3.0? This should be easy as I've done the merge not too long ago.

**Comment by xabbuh** [comment]
@fabpot sure, did so

**Pull request merged**